internal/app/grpcserver/server.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcserver
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net"
+
+	"connectrpc.com/conformance/internal/codec"
+	"connectrpc.com/conformance/internal/config"
+	conformancev1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+// Run runs the server according to server config read from the 'in' reader.
+func Run(_ context.Context, _ []string, inReader io.ReadCloser, outWriter io.WriteCloser, _ io.WriteCloser) error {
+	json := flag.Bool("json", false, "whether to use the JSON format for marshaling / unmarshaling messages")
+	host := flag.String("host", config.DefaultHost, "the host for the conformance server")
+	port := flag.String("port", config.DefaultPort, "the port for the conformance server ")
+
+	flag.Parse()
+
+	// Read the server config from  the in reader
+	data, err := io.ReadAll(inReader)
+	if err != nil {
+		return err
+	}
+
+	codec := codec.NewCodec(*json)
+
+	// Unmarshal into a ServerCompatRequest
+	req := &conformancev1alpha1.ServerCompatRequest{}
+	if err := codec.Unmarshal(data, req); err != nil {
+		return err
+	}
```

Comment:
Ditto about looking at #626. I changed `Codec` so its very different now, so this can read the message without using `io.ReadAll`.